### PR TITLE
wpcom-proxy-request: trigger request error when iframe is not loaded

### PR DESCRIPTION
**Pull request by jsnajdr**
Fixes an error frequently reported in Sentry where the `iframe.contentWindow` is `null` although the `wpcom-proxy-request` made sure that the `iframe` is created and loaded before it submits any requests to it.

The most likely cause is that some 3rd party script or extension is removing the `iframe`, effectively "stealing" it from the `wpcom-proxy-request` library.

Discussed also in this thread: https://github.com/Automattic/wp-calypso/pull/75138#issuecomment-1633819827

This is how the Sentry error looks like:

Screenshot 2023-07-25 at 10 40 51

**How to test:**
Load Calypso in dev mode, and after it's loaded, then in devtools find the proxy `iframe` element and delete it. Since that moment, you should be seeing errors in console every time the app wants to do a REST request:

Screenshot 2023-07-25 at 10 29 54

### CODE REVIEW

File: packages/wpcom-proxy-request/src/index.js
Comment (tyxla) [inline packages/wpcom-proxy-request/src/index.js]: Do we really need the `new` here? Seems like the error object is actually instantiated inside and returned by the function, so there's no need to make the `WPError` an actual instance too.

```suggestion
			WPError( { status_code: 500, error_description: 'proxy iframe element is not loaded' } ),
```

Comment (tyxla) [inline packages/wpcom-proxy-request/src/index.js]: Furthermore, I wonder if `500` is the right error type. This appears more like a client-side issue and not a server-side issue, so I expected a 4xx status code, like `406` or `405`. WDYT?

**Review by tyxla — COMMENTED** [review]
This looks good to me, thanks for addressing it @jsnajdr 👍 

I've left a couple of questions for your consideration before merging this, let me know what you think.

File: packages/wpcom-proxy-request/src/index.js
Comment (jsnajdr) [reply packages/wpcom-proxy-request/src/index.js]: Good catch, `WPError` is indeed not a class constructor, it's a factory function that creates `Error`s. Fixed.

Comment (jsnajdr) [reply packages/wpcom-proxy-request/src/index.js]: A client-side issue, reported as a 4xx error, means that there is something wrong with the request. Bad URL, missing information, some part is too large... But here the request is perfectly fine, we're just unable to process it. That's why I think 5xx is better. There are 5xx errors like 502 Bad Gateway or 504 Gateway Timeout that talk about a "gateway" going wrong. Maybe we could use 502 because that's exactly what we have here, a broken gateway (the proxy frame). But there's very little practical difference between 500 and 502.

Comment (tyxla) [reply packages/wpcom-proxy-request/src/index.js]: The thing is we're unable to process it on the client, not on the server, which is why I fail to see why we'd want to use a server error code. Anyway, I'm happy to leave the final decision to you.

**Review by tyxla — APPROVED** [review]
LGTM 🚀 

For the protocol, I'd personally prefer a client-side error code, since the failure is on the client and not the server, but that's just my personal preference and it's not enough to block the PR.

File: packages/wpcom-proxy-request/src/index.js
Comment (jsnajdr) [reply packages/wpcom-proxy-request/src/index.js]: In our case, the client is the app code that calls `wp.req.get()`. If there was some problem with the request's method, URL or headers, that would be a reason for a 4xx error.

The proxy iframe is not a client, it's a proxy. It runs in the browser, in the same tab (but different window and document) as the client code, but it's not _the_ client. It processes the requests originated in the client.

> I expected a 4xx status code, like 406 or 405.

Even if I wanted to use a 4xx code, I don't see any good one. 400 means "bad request", but our request is good. 406 means "not acceptable", i.e., the constraints specified in `Accept-*` headers can't be satisfied. Not our case. 405 means "method not allowed", like when the request is `PUT` but the server supports only `POST`. Also irrelevant.

On the other hand, there are good 5xx errors about gateways, like 502.

Comment (tyxla) [reply packages/wpcom-proxy-request/src/index.js]: Your 502 rationale makes sense 👍 Thanks for elaborating

**Comment by tyxla** [comment]
🤔 Okay, this makes sense and it provides a bit more information about where it occurred in the error stacktrace. That being said, I wonder if this is helpful enough to actually locate the underlying error? Seems to be coming from here:

https://github.com/Automattic/wp-calypso/blame/add/cli-progress-bar/client/landing/stepper/declarative-flow/internals/steps-repository/importer/index.tsx#L102 - cc @bogiii does it ring a bell?

**Comment by jsnajdr** [comment]
There's another report by Sentry now, and this time it's in the `<QuerySiteKeyrings>` component. Used by stats for Google My Business. Totally unrelated to the first report about importers.

So far it seems that we just replaced one error with another 🙂

**Pull request merged**
